File: Leetcode/Medium/701_insert_into_a_binary_search_tree.py
# Input: root = [40,20,60,10,30,50,70], val = 25
# Output: [40,20,60,10,30,50,70,null,null,25]

""" 
Input: root = [4,2,7,1,3,null,null,null,null,null,null], val = 5
Output: [4,2,7,1,3,5]
"""
import math
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return TreeNode(val=val)

        if val < root.val:
            root.left = self.insertIntoBST(root.left, val=val)
        else:
            root.right = self.insertIntoBST(root.right, val=val)

        return root


def showTree(root: Optional[TreeNode]) -> None:
    if not root:
        return None
    print(root.val)
    if root.left is not None:
        showTree(root.left)
    if root.right is not None:
        showTree(root.right)


def checkBST(root: Optional[TreeNode]) -> None:
    isValidBST(root, min=-math.inf, max=math.inf)


def isValidBST(root: Optional[TreeNode], min: float, max: float) -> None:
    if not root:
        return

    if root.val < min or root.val > max:
        raise ValueError(f"Invalid value {root.val}. ({min}, {max})")

    isValidBST(root.left, min, root.val - 1)
    isValidBST(root.right, root.val + 1, max)


if __name__ == "__main__":
    tree1 = TreeNode(
        val=40,
        left=TreeNode(
            val=20, left=TreeNode(val=10), right=TreeNode(val=30, left=TreeNode(val=19))
        ),
        right=TreeNode(val=60, left=TreeNode(val=50), right=TreeNode(val=70)),
    )

    solution = Solution()
    result = solution.insertIntoBST(tree1, val=25)
    showTree(result)
    checkBST(tree1)

File: Leetcode/Medium/701_insert_into_a_binary_search_tree.go
package main

import (
	"fmt"
	"log"
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func InorderTree(tree *TreeNode) {
	if tree == nil {
		return
	}

	InorderTree(tree.Left)
	fmt.Printf("%d,", tree.Val)
	InorderTree(tree.Right)
}

func IsValidBST(tree *TreeNode) bool {
	if tree == nil {
		return true
	}

	return checkBST(tree, math.Inf(-1), math.Inf(1))

}

func InsertBST(tree *TreeNode, val int) *TreeNode {
	if tree == nil {
		return &TreeNode{Val: val}
	}

	if val < tree.Val {
		tree.Left = InsertBST(tree.Left, val)
	} else {
		tree.Right = InsertBST(tree.Right, val)
	}

	return tree

}

func FindMinNode(tree *TreeNode) *TreeNode {
	cur := tree
	for cur.Left != nil {
		cur = cur.Left
	}
	return cur
}

func RemoveBST(tree *TreeNode, val int) *TreeNode {
	if tree == nil {
		return nil
	}

	if val < tree.Val {
		tree.Left = RemoveBST(tree.Left, val)
	} else if val > tree.Val {
		tree.Right = RemoveBST(tree.Right, val)
	} else {
		if tree.Left == nil {
			return tree.Right
		} else if tree.Right == nil {
			return tree.Left
		} else {
			minNode := FindMinNode(tree.Right)
			tree.Val = minNode.Val
			tree.Right = RemoveBST(tree.Right, minNode.Val)
		}

	}

	return tree

}

func checkBST(tree *TreeNode, min, max float64) bool {
	if tree == nil {
		return true
	}

	//fmt.Printf("Value: %v. Range [%v_%v]\n", tree.Val, min, max)
	if float64(tree.Val) < min || float64(tree.Val) > max {
		return false
	}
	left_result := checkBST(tree.Left, min, float64(tree.Val))
	if !left_result {
		return false
	}
	right_result := checkBST(tree.Right, float64(tree.Val), max)
	if !right_result {
		return false
	}

	return true

}

func main() {
	tree1 := &TreeNode{
		Val: 40,
		Left: &TreeNode{
			Val:   20,
			Left:  &TreeNode{Val: 10},
			Right: &TreeNode{Val: 30, Left: &TreeNode{Val: 21}},
		},
		Right: &TreeNode{
			Val: 60,
			Left: &TreeNode{
				Val: 50,
			},
			Right: &TreeNode{
				Val: 70,
			},
		},
	}
	isValid1 := IsValidBST(tree1)
	if !isValid1 {
		log.Fatalln("Is not valid BST")
	}

	InorderTree(tree1)
	fmt.Printf("\n")
	tree1 = InsertBST(tree1, 25)
	isValid2 := IsValidBST(tree1)
	if !isValid2 {
		log.Fatalln("Is not valid BST")
	}
	InorderTree(tree1)
	fmt.Printf("\n")
	tree1 = RemoveBST(tree1, 40)
	isValid3 := IsValidBST(tree1)
	if !isValid3 {
		log.Fatalln("Is not valid BST")
	}
	InorderTree(tree1)
	fmt.Printf("\n")
}
